test/compflags/ferguson/default-binary-name/sub_test

```diff
@@ -80,8 +80,11 @@ def run(source, exe_name, compopts, expect):
 
 run("test-default-binary-name.chpl", "test-default-binary-name", [], "Hello")
 run("test-default-binary-name.chpl", "test_exe", ["-o", "test_exe"], "Hello")
-run("test-default-binary-name-main-module.chpl", "M1", ["--main-module", "M1"], "M1 main")
-run("test-default-binary-name-main-module.chpl", "M2", ["--main-module", "M2"], "M2 main")
+run("test-default-binary-name-main-module.chpl", "test-default-binary-name-main-module", ["--main-module", "M1"], "M1 main")
+run("test-default-binary-name-main-module.chpl", "test-default-binary-name-main-module", ["--main-module", "M2"], "M2 main")
+run("test-default-binary-name2.chpl", "test-default-binary-name2", [], "Hi\nHello")
+run("test-default-binary-name3.chpl", "test-default-binary-name3", [], "Hi")
```

Comment:
The difference between
```
chpl test-default-binary-name3.chpl
```

and
```
chpl test-default-binary-name3.chpl M3.chpl
```

is surprising to me. It seems that if there is only one .chpl file present on the command line, Chapel considers that file the main module, even if `main` is defined in another module. I was imagining that the default "main module" would be "the module defining main". I think I've always thought it worked this way, but it does not.

As far as how to proceed I think:

 1. I'd recommend merging this PR as it's a good improvement that's been discussed. This potential misunderstanding is not new with this PR (the same test would have similar behavior on master).
 2. I think we need to either change this property of main module or document it better in the language specification. We could possibly consider renaming "main module" to "primary module" e.g. to reduce confusion.
 3. Other people involved in the discussion on the issue #8474 might appreciate a "heads up" about this.

Reply:
You're right that the question of identifying / defining the main module is independent of this PR and hasn't changed for some time.  I.e., we could change how the main module is defined or identified and this PR would still work and be documented correctly.  To describe how we got here:

Given that Chapel programs can just contain top-level code and, in that case, are given an implicit empty main, as in:

```chapel
writeln("Hello, world!");
```

we felt that it would be too confusing to say that if your simple Chapel program happened to `use` a module that defined a routine named `main()` that module would hijack the execution from you if it were not named on the command-line.  I.e., imagine I evolved the above program to:

```chapel
use SomeLib;
writeln("Hello, world!");
someLibCall();
```

and compiled it with:

```
chpl hello.chpl
```

If `SomeLib` ended up having a routine called `main()`, it'd be pretty surprising as a user to have the behavior of my program do something beyond printing hello world and making the obvious library call.

Now, arguably, a good library shouldn't define a `main()` and this is something we could check for, but that's how we got here.  We also rationalized it slightly by sometimes relying on a `main()` in a library module being used as a way to run standalone unit tests on that library, lacking another unit testing strategy and being reluctant at that time to add alternative ways of specifying the program's entry point (where some ideas included a `--main-function=unitTest()` flag or a well-defined entry point `proc unitTest()` that would be used instead of `proc main()` if compiling with a `--unitTest` flag).

Anyway, I don't disagree that there's the potential for confusion here, though I also don't think the current strategy is all that bad either (a potential piece of evidence being how long you've lived with Chapel without ever being bitten or surprised by it :) ).

Anyway, since we agree, I'll continue with this PR after making the documentation update you requested.

Reply:
The rationale for the current "main module" behavior sounds reasonable - I just think it is not clear from the language specification how nuanced it is. i.e. we need documentation improvement.